### PR TITLE
Feat/highlight demo

### PR DESCRIPTION
**Pull request by PLNech**
Highlighting helper + Demo showcasing three types of highlighting:
- `title` highlighted with default `toSpannableString`, -> **bold**
- `genre` highlighted with custom ForegroundColorSpan,  -> **yellow highlight**
- `actors` highlighted with full custom processing,  -> **capitalized** (see first hit `Mark FIte`

image

### CODE REVIEW

File: demo/src/main/kotlin/com/algolia/instantsearch/demo/highlighting/MovieAdapterHighlighted.kt [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    override fun onBindViewHolder(holder: MovieHighlightedViewHolder, position: Int) {
+        getItem(position)?.let { holder.bind(it) }
+    }
+
+    class MovieHighlightedViewHolder(private val view: View) : RecyclerView.ViewHolder(view) {
+
+        fun bind(movie: Movie) {
+            bindTitle(movie)
```
Comment (q-litzler) [inline demo/src/main/kotlin/com/algolia/instantsearch/demo/highlighting/MovieAdapterHighlighted.kt]: If you want to separate everything, I would go for a method signature resembling `fun Movie.bindTitleWith(view: TextView)`

File: helper/src/androidMain/kotlin/com/algolia/instantsearch/helper/highlighting/Extensions.kt [hunk 2]
```diff
@@ -0,0 +1,31 @@
+package com.algolia.instantsearch.helper.highlighting
```
Comment (q-litzler) [inline helper/src/androidMain/kotlin/com/algolia/instantsearch/helper/highlighting/Extensions.kt]: We should test these functions

File: helper/src/commonMain/kotlin/com/algolia/instantsearch/helper/highlighting/Highlightable.kt [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import com.algolia.search.model.Attribute
+import com.algolia.search.serialize.toHighlight
+import com.algolia.search.serialize.toHighlights
+import kotlinx.serialization.json.JsonObject
+
+
+/**
+ * Inheritors of this interface can use [getHighlight]/[getHighlights] methods to render highlights easily.
+ */
+interface Highlightable {
```
Comment (q-litzler) [inline helper/src/commonMain/kotlin/com/algolia/instantsearch/helper/highlighting/Highlightable.kt]: Let's make it public

File: helper/src/commonMain/kotlin/com/algolia/instantsearch/helper/highlighting/Highlightable.kt [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+    @Suppress("PropertyName") // Else implementers have to remember to specify @SerialName
+    val _highlightResult: JsonObject?
+
+    fun getHighlight(
```
Comment (q-litzler) [inline helper/src/commonMain/kotlin/com/algolia/instantsearch/helper/highlighting/Highlightable.kt]: Explicitly public, for the one below as well

File: helper/src/commonMain/kotlin/com/algolia/instantsearch/helper/highlighting/Highlightable.kt [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        preTag: String = DefaultPreTag,
+        postTag: String = DefaultPostTag
+    ): HighlightedString? = getHighlight(key.raw, findHighlight)
+
+    fun getHighlight(
+        key: String,
+        findHighlight: (JsonObject) -> JsonObject = { it },
+        preTag: String = DefaultPreTag,
+        postTag: String = DefaultPostTag
+    ): HighlightedString? = _highlightResult?.let { findHighlight(it).toHighlight(key) }?.let {
```
Comment (q-litzler) [inline helper/src/commonMain/kotlin/com/algolia/instantsearch/helper/highlighting/Highlightable.kt]: We should not use implicit function returns in public parts of the library. Let's have the explicitly typed and non inlined return instead.

File: helper/src/androidMain/kotlin/com/algolia/instantsearch/helper/highlighting/Extensions.kt [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    parts.forEach { (part, isHighlighted) ->
+        if (isHighlighted) inSpans(span) { append(part) }
+        else append(part)
+    }
+}
+
+public fun List<HighlightedString>.toSpannedString(
+    span: ParcelableSpan = StyleSpan(Typeface.BOLD)
+): SpannedString {
+    return buildSpannedString {
```
Comment (q-litzler) [inline helper/src/androidMain/kotlin/com/algolia/instantsearch/helper/highlighting/Extensions.kt]: We could simplify it with something like this:

```kotlin
    return buildSpannedString {
        this@toSpannedString.map { it.toSpannedString(span) }.forEachIndexed { index, span ->
            if (index > 0) append(", ")
            append(span)
        }
    }
```

**Comment by q-litzler** [comment]
See https://github.com/algolia/instantsearch-kotlin/pull/41 for cosmetic changes in TestHighlightab

File: core/src/commonMain/kotlin/com/algolia/instantsearch/core/highlighting/HighlightToken.kt [hunk 7]
```diff
@@ -4,4 +4,6 @@ package com.algolia.instantsearch.core.highlighting
 /**
  * A token from a highlighted string.
  */
-public data class HighlightToken(val content: String, val highlighted: Boolean = false)
+public data class HighlightToken(val content: String, val highlighted: Boolean = false) {
+    override fun toString(): String = if (highlighted) "_${content}_" else content
```
[the remark this answers is not in the document]
Comment (PLNech) [reply core/src/commonMain/kotlin/com/algolia/instantsearch/core/highlighting/HighlightToken.kt]: Yep, was quite practical when manipulating `HighlightedString`, compare:
- previous _toString_ `HighlightedString(original=foo<em>bar</em>baz, parts=[HighlightToken(content=foo, highlighted=false), HighlightToken(content=bar, highlighted=true), HighlightToken(content=baz, highlighted=false)])
`, which was a struggle to parse
- current _toString_: `HighlightedString(original=foo<em>bar</em>baz, parts=[foo, _bar_, baz])`

File: demo/src/main/kotlin/com/algolia/instantsearch/demo/highlighting/MovieAdapterHighlighted.kt (hunk 1 above)
Comment (PLNech) [reply demo/src/main/kotlin/com/algolia/instantsearch/demo/highlighting/MovieAdapterHighlighted.kt]: It feels weird to me to have on the model functions for for UI databinding, why do you feel it's best than in the `ViewHolder` responsible of the binding? :thinking:

File: helper/src/commonMain/kotlin/com/algolia/instantsearch/helper/highlighting/Highlightable.kt (hunk 5 above)
Comment (PLNech) [reply helper/src/commonMain/kotlin/com/algolia/instantsearch/helper/highlighting/Highlightable.kt]: From our chat: Inlined return acceptable as long as the type is explicit

File: demo/src/main/kotlin/com/algolia/instantsearch/demo/highlighting/MovieAdapterHighlighted.kt (hunk 1 above)
Comment (PLNech) [reply demo/src/main/kotlin/com/algolia/instantsearch/demo/highlighting/MovieAdapterHighlighted.kt]: Then again it's just a matter of what example we want to show our users, we can also inline them all in `bind` if you find it clearer:
```kotlin
fun bind(movie: Movie) {
            view.itemTitle.text = TextUtils.concat(movie.highlightedTitle?.toSpannedString(), " (${movie.year})")
            view.itemSubtitle.text = movie.highlightedGenres?.toSpannedString(BackgroundColorSpan(Color.YELLOW))
                ?: buildSpannedString { italic { append("unknown genre") } }
            view.itemFooter.text = movie.highlightedActors?.let { list ->
                list.sortedByDescending { it.highlightedTokens.size }.joinToString { highlight ->
                    highlight.tokens.joinToString("") {
                        if (it.highlighted) it.content.toUpperCase() else it.content
                    }
                }
            }
            Glide.with(view)
                .load(movie.image).placeholder(android.R.drawable.ic_media_play)
                .centerCrop()
                .into(view.itemImage)
        }
```

File: core/src/commonMain/kotlin/com/algolia/instantsearch/core/highlighting/HighlightToken.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (q-litzler) [reply core/src/commonMain/kotlin/com/algolia/instantsearch/core/highlighting/HighlightToken.kt]: Indeed. If that's the only reason, I don't think we should expose it as part of the public API. I think we have the same in FilterState, so maybe we can add a task to remove all of them later.

File: helper/src/androidMain/kotlin/com/algolia/instantsearch/helper/highlighting/Extensions.kt (hunk 6 above)
Comment (PLNech) [reply helper/src/androidMain/kotlin/com/algolia/instantsearch/helper/highlighting/Extensions.kt]: Good call!

**Comment by PLNech** [comment]
@q-litzler feedback addressed, #41 merged, let me know if it looks good to you :)

**Pull request merged**